### ISSUE TITLE
[native-mt] Darwin single thread dispatchers leak ObjC autorelease references

### ISSUE THREAD

**Issue by andersio** [Kotlin/kotlinx.coroutines#2322]
Kotlin: 1.4.10
Coroutines: 1.3.9-native-mt-2

Also filed as https://youtrack.jetbrains.com/issue/KT-42822 for the K/N runtime.

KT-42822 is about K/N `Worker`s not wrapping its work items with an autorelease pool. So when Kotlin code calls into ObjC/Swift on a `Worker`, any `autorelease` reference leaks unconditionally, e.g., any `id` or `NSObject*` return values.

However, even if KT-42822 is fixed, a fix at the coroutines library is still necessary, since kotlinx-coroutines-core runs its own EventLoop over the one in `Worker`.

The native single thread `CoroutineDispatcher` wraps the entire `EventLoop` with an autorelease pool, but not individual blocks as they get executed. This means autorelease references in these blocks would be pushed to a pool that is never drained until the `EventLoop` is closed.

So in other words, autorelease references are effectively leaked on any long living dispatchers created by `newSingleThreadContext()`.

https://github.com/Kotlin/kotlinx.coroutines/blob/1b0e196bc557f03fd9e13b81edb6997bddeb38e7/kotlinx-coroutines-core/native/src/Workers.kt#L48-L50

This could be solved by either:

* `native-mt` single-thread dispatchers wrap individual blocks with `autoreleasepool` before executing them; or
* wrap `processNextEvent()` with `autoreleasepool {}` maybe?

   https://github.com/Kotlin/kotlinx.coroutines/blob/1b0e196bc557f03fd9e13b81edb6997bddeb38e7/kotlinx-coroutines-core/native/src/Builders.kt#L77-L83

**Comment by slavonnet** [comment]
maybe native have some issue and park thread forever?

https://github.com/Kotlin/kotlinx.coroutines/pull/2374

**Comment by qwwdfsad** [comment]
Fixed in 1.4.3-native-mt
